### PR TITLE
[Merged by Bors] - feat: port AlgebraicGeometry.StructureSheaf

### PR DESCRIPTION
**Pull request by jjaassoonn**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/StructureSheaf.lean [hunk 1]
```diff
@@ -0,0 +1,1224 @@
...
+set_option linter.uppercaseLean3 false in
+#align algebraic_geometry.structure_presheaf_in_CommRing AlgebraicGeometry.structurePresheafInCommRing
+
+/-- Some glue, verifying that that structure presheaf valued in `CommRingCat` agrees
+with the `Type` valued structure presheaf.
+-/
+def structurePresheafCompForget :
+    structurePresheafInCommRing R ⋙ forget CommRingCat ≅ (structureSheafInType R).1 :=
+  NatIso.ofComponents (fun U => Iso.refl _) (fun i => by aesop)
+set_option linter.uppercaseLean3 false in#align algebraic_geometry.structure_presheaf_comp_forget AlgebraicGeometry.structurePresheafCompForget
```
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/StructureSheaf.lean:259]:
```suggestion
set_option linter.uppercaseLean3 false in
#align algebraic_geometry.structure_presheaf_comp_forget AlgebraicGeometry.structurePresheafCompForget
```

File: Mathlib/AlgebraicGeometry/StructureSheaf.lean [hunk 2]
```diff
@@ -0,0 +1,1224 @@
...
+instance isIso_to_global : IsIso (toOpen R ⊤) := by
+  let hom := CommRingCat.ofHom (algebraMap R (Localization.Away (1 : R)))
+  haveI : IsIso hom :=
+    IsIso.of_iso (IsLocalization.atOne R (Localization.Away (1 : R))).toRingEquiv.toCommRingCatIso
+  rw [to_global_factors R]
+  infer_instance
+#align algebraic_geometry.structure_sheaf.is_iso_to_global AlgebraicGeometry.StructureSheaf.isIso_to_global
+
+/-- The ring isomorphism between the ring `R` and the global sections `Γ(X, 𝒪ₓ)`. -/
+-- Porting note : was @[simps (config := { rhsMd := Tactic.Transparency.semireducible })]
```
Comment (kbuzzard) [inline Mathlib/AlgebraicGeometry/StructureSheaf.lean:1007]: Are you clear about what is happening here? I'm not, so basically I'm saying "I'm reviewing your PR but this line I am not capable of reviewing properly".

**Review by kbuzzard — APPROVED** [review]
This is great to see -- a complex file like this in the algebraic geometry part of mathlib, and the porting process looks to me to be relatively straightforward. Do you understand why you had to add so many `PrimeSpectrum.`? e.g. Lean 3 `rw [mem_zero_locus,...]` becomes Lean 4 `rw [PrimeSpectrum.mem_zeroLocus, ...]` on line 671 (but there are many other examples). Edit: OK I understand this now; the issue is that `open PrimeSpectrum` is opening `AlgebraicGeometry.PrimeSpectrum` rather than what would in Lean 3 have been called `_root_.PrimeSpectrum`. I asked about this [here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/open.20.60bar.60.20in.20namespace.20.60foo.60.20when.20.60foo.2Ebar.60.20exists/near/362537281).

 I am not capable of reviewing the `simps!` line which I flagged because I don't understand what the autoporter is suggesting. Other than that, this looks fine; my only comment are superficial.

File: Mathlib/AlgebraicGeometry/StructureSheaf.lean (hunk 2 above)
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/StructureSheaf.lean:1008]: I am not sure what to do with this, but I see that `@[simps (config := { rhsMd := semireducible })]` in other ported file are just replaced with `simps!`

**Comment by Ruben-VandeVelde** [comment]
@jjaassoonn can this be marked as awaiting-review?

**Comment by jjaassoonn** [comment]
> @jjaassoonn can this be marked as awaiting-review?

Ah yes. Thanks for the reminder

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Comment by fpvandoorn** [comment]
Oh, I see that @Parcly-Taxel merged master. Note that if you do that after `bors merge`, it will cancel the batch that the PR is in.

**Pull request closed without merging**